sinks/api/types.go

```diff
@@ -0,0 +1,167 @@
+	// An int64 value.
+	ValueInt64 MetricValueType = iota
+	// A boolean value
+	ValueBool
+	// A double-precision floating point number.
+	ValueDouble
+)
+
+func (self MetricValueType) String() string {
+	switch self {
+	case ValueInt64:
+		return "int64"
+	case ValueBool:
+		return "bool"
+	case ValueDouble:
+		return "double"
+	}
+	return ""
+}
+
+type MetricUnitsType int
+
+const (
+	// A metric with no units.
+	UnitsNone MetricUnitsType = iota
+	// A metric in bytes.
+	UnitsBytes
+	// A metric in milliseconds.
+	UnitsMilliseconds
+	// A metric in nanoseconds.
+	UnitsNanoseconds
+)
+
+func (self *MetricUnitsType) String() string {
+	switch *self {
+	case UnitsBytes:
+		return "bytes"
+	case UnitsMilliseconds:
+		return "ms"
+	case UnitsNanoseconds:
+		return "ns"
+	}
+	return ""
+}
+
+type LabelDescriptor struct {
+	// Key to use for the label.
+	Key string `json:"key,omitempty"`
+
+	// Description of the label.
+	Description string `json:"description,omitempty"`
+}
+
+type MetricDescriptor struct {
+	// The unique name of the metric.
+	Name string
+
+	// Description of the metric.
+	Description string
+
+	// Descriptor of the labels used by this metric.
+	Labels []LabelDescriptor
+
+	// Type and value of metric data.
+	Type      MetricType
+	ValueType MetricValueType
+	Units     MetricUnitsType
+}
+
+type Metric struct {
+	// The name of the metric. Must match an existing descriptor.
+	Name string
+
+	// The labels and values for the metric. The keys must match those in the descriptor.
+	Labels map[string]string
+
+	// The start and end time for which this data is representative.
+	Start time.Time
+	End   time.Time
+
+	// The value of the metric. Must match the type in the descriptor.
+	Value interface{}
+}
+
+// SupportedStatMetric represents a resource usage stat metric.
+type SupportedStatMetric struct {
+	MetricDescriptor
+
+	// Returns whether this metric is present.
+	HasValue func(*cadvisor.ContainerSpec) bool
+
+	// Returns the desired data point for this metric from the stats.
+	GetValue func(*cadvisor.ContainerSpec, *cadvisor.ContainerStats) interface{}
+}
+
+// Timeseries represents a single metric.
+type Timeseries struct {
+	Metric           *Metric
+	MetricDescriptor *MetricDescriptor
+}
+
+// CompositeData represents different kinds of data that needs to be stored.
+type CompositeData struct {
+	Metrics []Timeseries
+	// TODO: Add map[string]string to support storing raw data like node or pod status, spec, etc.
+	// TODO: Add events.
+}
+
+// Sink is the interface that needs to be implemented by all storage backends.
+type Sink interface {
+	// Stores input CompositeData into the backend.
+	Store(*CompositeData) error
+
+	// Returns debuf information specific to the sink.
+	DebugData() string
+}
+
+// Codec represents an engine that translated from sources.api to sink.api objects.
+type Decoder interface {
+	// Metrics returns the metrics found in input as a timeseries slice.
+	Metrics(input interface{}) ([]Timeseries, error)
```

Comment:
Stronger interfaces make it easier for the implementer and for the person writing the code. With an interface I have no idea what types of inputs I get and what I need to support. A strong type tells me exactly what that is and I can choose what to do if I can't handle it.